### ISSUE TITLE
Build failures with gcc-14

### ISSUE THREAD

**Issue by hainest** [dyninst/dyninst#1740]
The Annotation class runs into the new calloc-transposed-args warning.

```console
In file included from /dyninst/src/common/src/Annotatable.C:35:
/dyninst/src/common/h/Annotatable.h: In member function 'bool Dyninst::AnnotatableDense::addAnnotation(const void*, Dyninst::AnnotationClassID)':
/dyninst/src/common/h/Annotatable.h:175:68: error: 'void* calloc(size_t, size_t)' sizes specified with 'sizeof' in the earlier argument and not in the later argument [-Werror=calloc-transposed-args]
  175 |                         annotations->data = (anno_list_t *) calloc(sizeof(anno_list_t), (size));
      |                                                                    ^~~~~~~~~~~~~~~~~~~
```

This is definitely a bug. Moreover, it looks like the only reason this code ever worked is because it allocates enough space for a pointer which is all the `annotations` list ever holds for all of the places it's used in Dyninst. It also looks like the only way these annotation classes are ever used is to store a single object for some class that can be fetched by name even though the annotation cache can store many such objects. I've not fully convinced myself that these annotations can just be removed altogether and replaces with class member variables, but I would recommennd moving in that direction, given that the annotations are clearly fundamentally broken.

---

List of annotation classes:

FieldUpPtrAnno
ImageVariableUpPtrAnno
Stack_Anno_Block_Effects
Stack_Anno_Call_Effects
Stack_Anno_Insn_Effects
Stack_Anno_Intervals
SymbolFileNameAnno
SymbolVersionNamesAnno
TypeUpPtr
UserRegionsAnno
UserTypesAnno

**Comment by hainest** [comment]
This one's from gcc-13 on Fedora-39 using `-DENABLE_STATIC_LIBS=ON`.

```
In file included from /__w/dyninst/dyninst/src/symtabAPI/h/Type.h:38,
                 from /__w/dyninst/dyninst/src/dyninstAPI/h/BPatch_type.h:39,
                 from /__w/dyninst/dyninst/src/dyninstAPI/src/BPatch_collections.h:35,
                 from /__w/dyninst/dyninst/src/dyninstAPI/src/BPatch_type.C:38:
In member function 'bool Dyninst::AnnotatableDense::addAnnotation(const void*, Dyninst::AnnotationClassID)',
    inlined from 'bool Dyninst::AnnotatableDense::addAnnotation(const T*, Dyninst::AnnotationClass<T>&) [with T = BPatch_field]' at /__w/dyninst/dyninst/src/common/h/Annotatable.h:248:28,
    inlined from 'BPatch_field::BPatch_field(Dyninst::SymtabAPI::Field*, BPatch_dataClass, int, int)' at /__w/dyninst/dyninst/src/dyninstAPI/src/BPatch_type.C:329:26,
    inlined from 'void* Dyninst::AnnotationClass<T>::allocate() [with T = BPatch_field]' at /__w/dyninst/dyninst/src/common/h/Annotatable.h:104:21:
/__w/dyninst/dyninst/src/common/h/Annotatable.h:164:15: error: array subscript 0 is outside array bounds of 'Dyninst::AnnotatableDense [0]' [-Werror=array-bounds=]
  164 |          if (!annotations)
      |               ^~~~~~~~~~~
In member function 'void* Dyninst::AnnotationClass<T>::allocate() [with T = BPatch_field]':
cc1plus: note: source object is likely at address zero
In member function 'bool Dyninst::AnnotatableDense::addAnnotation(const void*, Dyninst::AnnotationClassID)',
    inlined from 'bool Dyninst::AnnotatableDense::addAnnotation(const T*, Dyninst::AnnotationClass<T>&) [with T = BPatch_field]' at /__w/dyninst/dyninst/src/common/h/Annotatable.h:248:28,
    inlined from 'BPatch_field::BPatch_field(Dyninst::SymtabAPI::Field*, BPatch_dataClass, int, int)' at /__w/dyninst/dyninst/src/dyninstAPI/src/BPatch_type.C:329:26,
    inlined from 'void* Dyninst::AnnotationClass<T>::allocate() [with T = BPatch_field]' at /__w/dyninst/dyninst/src/common/h/Annotatable.h:104:21:
/__w/dyninst/dyninst/src/common/h/Annotatable.h:166:25: error: array subscript 0 is outside array bounds of 'Dyninst::AnnotatableDense [0]' [-Werror=array-bounds=]
  166 |             annotations = (aInfo *) malloc(sizeof(aInfo));
      |             ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by hainest** [comment]
Fixed via #1788
